gcsweb/cmd/gcsweb/gcsweb.go

```diff
@@ -0,0 +1,372 @@
+const gcsPath = "/gcs" // path for GCS browsing on this server
+
+var flPort = flag.Int("p", 8080, "port number on which to listen")
+var flIcons = flag.String("i", "/icons", "path to the icons directory")
+var flVersion = flag.Bool("version", false, "print version and exit")
+
+const (
+	iconFile = "/icons/file.png"
+	iconDir  = "/icons/dir.png"
+	iconBack = "/icons/back.png"
+)
+
+type strslice []string
+
+// String prints the strlice as a string.
+func (ss *strslice) String() string {
+	return fmt.Sprintf("%v", *ss)
+}
+
+// Set appends a value onto the strslice.
+func (ss *strslice) Set(value string) error {
+	*ss = append(*ss, value)
+	return nil
+}
+
+// Only buckets in this list will be served.
+var allowedBuckets strslice
+
+func main() {
+	flag.Var(&allowedBuckets, "b", "GCS bucket to serve (may be specified more than once)")
+	flag.Parse()
+
+	if *flVersion {
+		fmt.Println(version.VERSION)
+		os.Exit(0)
+	}
+
+	log.Printf("starting")
+	rand.Seed(time.Now().UTC().UnixNano())
+
+	// Canonicalize allowed buckets.
+	for i := range allowedBuckets {
+		allowedBuckets[i] = joinPath(gcsPath, allowedBuckets[i])
+		log.Printf("allowing %s", allowedBuckets[i])
+		http.HandleFunc(allowedBuckets[i]+"/", gcsRequest)
+		http.HandleFunc(allowedBuckets[i], func(w http.ResponseWriter, r *http.Request) {
+			http.Redirect(w, r, allowedBuckets[i]+"/", http.StatusMovedPermanently)
+		})
+	}
+
+	// Serve icons.
+	http.Handle("/icons/", http.StripPrefix("/icons/", http.FileServer(http.Dir(*flIcons))))
+
+	// Serve HTTP.
+	http.HandleFunc("/", otherRequest)
+	log.Printf("serving on port %d", *flPort)
+	log.Fatal(http.ListenAndServe(fmt.Sprintf(":%d", *flPort), nil))
+}
+
+func otherRequest(w http.ResponseWriter, r *http.Request) {
+	newTxnLogger(r)
+	http.NotFound(w, r)
+}
+
+func gcsRequest(w http.ResponseWriter, r *http.Request) {
+	logger := newTxnLogger(r)
+
+	if r.Method != "GET" {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	// e.g. "/gcs/bucket/path/to/object" -> "/bucket/path/to/object"
+	path := strings.TrimPrefix(r.URL.Path, gcsPath)
+	// e.g. "/bucket/path/to/object" -> ["bucket", "path/to/object"]
+	bucket, object := splitBucketObject(path)
+
+	url := joinPath(gcsBaseURL, bucket)
+	url += "?delimiter=/"
+
+	if object != "" {
+		// Adding the last slash forces the server to give me a clue about
+		// whether the object is a file or a dir.  If it is a dir, the
+		// contents will include a record for itself.  If it is a file it
+		// will not.
+		url += "&prefix=" + object + "/"
+	}
+
+	if markers, found := r.URL.Query()["marker"]; found {
+		url += "&marker=" + markers[0]
+	}
+
+	resp, err := http.Get(url)
+	if err != nil {
+		logger.Printf("GET %s: %s", url, err)
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "http.Get: %v", err)
+		return
+	}
+	defer resp.Body.Close()
+
+	logger.Printf("GET %s: %s", url, resp.Status)
+
+	if resp.StatusCode != http.StatusOK {
```

Comment:
The GCS documentation suggests retrying 500s with exponential backoff. Why not do that here?